### DISCUSSION TITLE
🐛 em.upsert() does not fetch the primary key

### DISCUSSION THREAD

**Discussion by lukas-runge** [mikro-orm/mikro-orm#6164] [Q&A] [answered]
### Describe the bug

Hi everyone 👋,

after trying for days to get upsert to work I am finally pulling the white flag. I am always confronted with this error: `NotNullConstraintViolationException`. I do not understand why this is happening as the docs say that MikroORM should automatically fetch the primary key when it's missing in `data`.

### Reproduction

Please have a look at the minimal reproduction repo here: https://github.com/lukas-runge/mikroorm-upsert-reproduction 🙌

### What driver are you using?

@mikro-orm/sqlite

### MikroORM version

6.3.13

### Node.js version

22.9.0

### Operating system

macOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment, marked as answer]
There are two issues with your repro:
- upserting a relation is not supported, that's why you see the JSON in your query
- upserting without a PK requires you to have a db default for it, which you don't have

For the first part, that is a wontfix, you cant upsert the address since it has no unique property, it would always result in an insert query (even if supported).

For the second part, its a wontfix too if you want to use UUID v7, since that is currently not supported in postgres natively afaik. If you would use v4, you can do this:

```ts
@PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
id!: string;
```

Alternatively, instead of upserting via POJO (what you do), you can upsert an entity instance - you first create it, which will execute the property initializer:

```ts

const user = em.create(User, {
	firstName: "John",
	lastName: "Doe",
	email: "johndoe@london.co.uk",
	address: {
		street: "Baker Street 221B",
		city: "London"
	}
});

await em.upsert(user);
console.log(user.id);
```

This solution won't require the `defaultRaw` as suggested above. It will also only upsert the user and leave out the address, since again, upserting relations is not supported.

**Comment by lukas-runge** [comment]
@B4nan Ok - thank you very much for taking the time to clarify that for me! 🙏
Maybe a notice on that in the docs would be helpful to other people encountering this in the future.

Have a great week! 🙌

Kind regards,
@lukas-runge
